portfolio/src/main/client/src/util/html.ts

```diff
@@ -0,0 +1,153 @@
+import type {Arrunk} from './types';
+
+type InterpolationValue = Arrunk<string | Node | null | undefined> | EventListener;
+
+/**
+ * Helper function to iterate over all descendants of a node.
+ * @param node The node to iterate over.
+ */
+function* descendants(node: Node): Generator<Node> {
+  for (const child of node.childNodes) {
+    yield child;
+    for (const grandchild of descendants(child)) {
+      yield grandchild;
+    }
+  }
+}
+
+/**
+ * @param arr The array to flatten.
+ * @returns The flattened array.
+ */
+function flatten<T>(arr: Arrunk<T>[]): T[] {
+  const newArr = [];
+  for (const item of arr) {
+    if (Array.isArray(item)) {
+      newArr.push(...item);
+    } else {
+      newArr.push(item);
+    }
+  }
+  return newArr;
+}
+
+// eslint-disable-next-line valid-jsdoc
+/**
+ * Helper function to parse HTML into multiple DOM nodes. Meant to be used
+ * as a template literal tag:
+ * ```
+ * const myElem = htmls`<td>this is the html</td><td>this is the html</td>`;
+ * ```
+ *
+ * You can interpolate HTML elements into the template literal, and it will
+ * insert them properly:
+ * ```
+ * const btn = document.createElement('button');
+ * btn.addEventListener(() => ...);
+ * const myElem = htmls`<p>some text<p>${btn}<p>with a button inside</p>`;
+ * ```
+ *
+ * You can also interpolate arrays of HTML elements or strings.
+ *
+ * Finally, you can interpolate functions, but only to use them as event handlers
+ * in @-attributes. Using them anywhere else is undefined behaviour.
+ * ```
+ * const myHandler = () => console.log('hi');
+ * const myElem = htmls`<button @click='${myHandler}'>Say hi</button>`;
```

Comment:
You shouldn't be adding a templating language here as this reduces readability. In general, it's preferred not to have "magic" that changes the syntax unless you're writing a general library. It should be trivial to rewrite lines like this as:
```
const myElem = htmls`<button>Say hi</button>`;
myElem.addEventListener('click', () => console.log('hi'));
```

This will also simplify the code in this file.